middleware/grpc_logging.go

```diff
@@ -12,15 +12,26 @@ import (
 
 const gRPC = "gRPC"
 
-// ServerLoggingInterceptor logs gRPC requests, errors and latency.
-var ServerLoggingInterceptor = func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
-	begin := time.Now()
-	resp, err := handler(ctx, req)
-	entry := logging.With(ctx).WithFields(log.Fields{"method": info.FullMethod, "duration": time.Since(begin)})
-	if err != nil {
-		entry.WithField("request", req).WithError(err).Warn(gRPC)
-	} else {
-		entry.Debugf("%s (success)", gRPC)
+// GRPCServerLog logs grpc requests, errors, and latency.
+type GRPCServerLog struct {
+	// WithRequest will log the entire request rather than just the error
+	WithRequest bool
+}
+
+// UnaryServerInterceptor returns an interceptor that logs gRPC requests
+func (s GRPCServerLog) UnaryServerInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
```

Comment:
You don't have to do this as `return func` - just declare a method with the right signature and Go will create a closure for it if you refer to it as `serverLog.UnaryServerInterceptor`
https://golang.org/ref/spec#Method_values

Reply:
That's neat!